modules/tensor_mechanics/src/userobjects/nodal_patch_recovery/NPRRankTwoAux.C

```diff
@@ -0,0 +1,44 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "NPRRankTwoAux.h"
+
+registerMooseObject("TensorMechanicsApp", NPRRankTwoAux);
+
+InputParameters
+NPRRankTwoAux::validParams()
```

Comment:
I don't get it... why is this called `Aux` when it's inheriting from a user object?

Ok you are concerned about having the user object loop over the evaluable element range ... but can you tell me why? Are you doing DG in your tensor mechanics computations? Right now it seems to me like you are going to be leveraging off-process material property information in your computation of *local* user object data, and this should now work (well after we get the libmesh links worked out). Are you planning on using *nonlocal* user object data somewhere? If you are, then we should think about what communication we want to do in `NPRBase::finalize`.

Reply:
Yeah I'll think about a better name.

In `NPRAux.C`, for each node, I look at its connected elements. Therefore, the user object needs to have data from algebraically ghosted elements as well. That's my motivation of running user objects over evaluable elements.

Next, in that userobject, to compute the data on each element, I need to compute the requested material properties. That's my motivation of running materials over evaluable elements.

Now, on a second thought. I think I get away with all of these by doing the proper `finalize` in the user object, and keep the material loop and the user object loop still on local elements instead of evaluable elements. In `finalize`, I just need to loop over evaluable elements, and if the element is nonlocal, I can do the proper communication to grab the data from other processors.

Am I understanding it correctly now?

Reply:
> In `NPRAux.C`, for each node, I look at its connected elements. Therefore, the user object needs to have data from algebraically ghosted elements as well. That's my motivation of running user objects over evaluable elements.

We need to delineate consumption and production. I understand that you need the things your user object consumes (stateful material properties) to be computed on algebraically ghosted elements and hence why it makes sense to me to run the `ComputeMaterialsObjectThread` on evaluable elements ... this will ensure that the data your user object *consumes* are complete.

However, we must now turn to what your user object *produces*. Does *that* need to computed for algebraically ghosted elements? Are you going to have objects that need to get user object data from algebraically ghosted elements?

Reply:
My (perhaps very poor) conceptual understanding is that your nodal patch recovery user object is going to *produce* better *local* information by considering/consuming nonlocal information. But if you are doing something like DG, which will need to consume nonlocal data, then there I would understand the need to make sure your user object *produces* nonlocal data in addition to *consuming*  nonlocal data.

Reply:
This is what my userobject produces:
```cpp
/// The element-level A matrix
std::vector<DenseMatrix<Real>> _Ae;
/// The element-level b vector
std::vector<DenseVector<Real>> _be;
virtual Real nodalPatchRecovery(const Point &, const std::vector<dof_id_type> &) const;
```
`_Ae` and `_be` are matrices and vectors for each element. The `NPRAux` will call `nodalPatchRecovery` with a point and a patch of elements. Some of the elements will be off-processor.

This is what my userobject consumes:
```
    be.add(computeValue(), p);
```
where `computeValue()` may need (stateful)material properties. See above, because `_be` should be available on some off-processor elements, _I thought_ what this user object consumes should also be available on those off-processor elements.

Reply:
> are there things that are going to want the user object produced data for elements that are not local?

Yes, the `NPRAux` will want to call `NPRBase::nodalPatchRecovery` with elements that are not local.

Reply:
Ok yes I see. We'll definitely need the communication then, and I think pull will be the way to go

Reply:
@dschwen Regarding the naming of this userobject: I completely agree with @lindsayad that calling a userobject `NPRRankTwoAux` is very confusing. 

I chose this name because I was thinking ahead about making actions. I think simply prepending `NPR` in front of `RankTwoAux` will make our lives easier when incorporating nodal patch recovery in TM action. But perhaps I should also add a suffix to tell people this is _NOT_ an auxkernel. Maybe `NPRRankTwoAuxUserObject`? What do you think?

Reply:
`NPRRankTwoAuxUserObject` seems fine. Just a reminder though that we do have the `IndexableProperty` class available in the framework  now, which takes care of all teh indexing parameters and indexing and would make it trivial for this object to support all sorts of tensors (inclusding the upcoming Mandel notation tensors!).

Reply:
The other option is to only provide this UO for Real material properties and have the user/action add a separate material that extracts the component (we do already have that). I don't like that option quite as much as my IndexableProperty solution (because it adds more clutter to the inputs if actions are not used), but in either case there should not be special objects to deal with Rank*Tensor components anymore.

Reply:
Okay, good to know that `IndexableProperty` has made its way in! I'll try to use that.

BTW, what's a Mandel notation tensor?